File: presentation/src/main/java/com/octo/presentation/CityForecastPresenterImpl.java
package com.octo.presentation;

import android.content.res.Resources;

import com.octo.usecases.CityForecastPresenter;
import com.octo.usecases.Forecast;

import java.text.DateFormat;
import java.text.DecimalFormat;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class CityForecastPresenterImpl implements CityForecastPresenter {
    private static final String NORMAL_DATE_FORMAT = "EEEE 'à' H'h'";
    private static final String TODAY_DATE_FORMAT = "'aujourd''hui à ' H'h'";
    private final DateFormat normalDateFormat = new SimpleDateFormat(NORMAL_DATE_FORMAT, Locale.getDefault());
    private final DateFormat todayDateFormat = new SimpleDateFormat(TODAY_DATE_FORMAT, Locale.getDefault());
    private final CityForecastView view;
    private final Resources resources;
    private final DateUtilsInjector dateUtilsInjector;

    public CityForecastPresenterImpl(final CityForecastView view, final Resources resources, final DateUtilsInjector dateUtilsInjector) {
        this.dateUtilsInjector = dateUtilsInjector;
        this.view = view;
        this.resources = resources;
    }

    @Override
    public void onEmptyInput() {
        view.displayMessage(resources.getString(R.string.empty_input));
    }

    @Override
    public void onUnknownCity() {
        view.displayMessage(resources.getString(R.string.unknown_city));
    }

    @Override
    public void onGenericException() {
        view.displayMessage(resources.getString(R.string.unavailable_forecasts));
    }

    @Override
    public void onUnavailableForecasts() {
        view.displayMessage(resources.getString(R.string.unavailable_forecasts));
    }

    @Override
    public void onForecasts(String city, Forecast worstForecast, Forecast bestForecast) {
        view.displayViewModel(ForecastViewModel.builder()
            .setTitle(resources.getString(R.string.city_forecast, city))
            .setWorstTemperatureLabel(getTemperatureLabel(worstForecast.getTemperature()))
            .setBestTemperatureLabel(getTemperatureLabel(bestForecast.getTemperature()))
            .setWorstForecastLabel(getForecastLabel(worstForecast))
            .setBestForecastLabel(getForecastLabel(bestForecast))
            .build());
    }

    private String getForecastLabel(Forecast forecast) {
        final Date date = forecast.getDate();
        final String dateFormatted = getDateFormat(date).format(date);
        return capitalize(dateFormatted);
    }

    private String getTemperatureLabel(final Float temperature) {
        final DecimalFormat decimalFormat = new DecimalFormat();
        decimalFormat.setMaximumFractionDigits(0);
        return resources.getString(R.string.temperature_label, decimalFormat.format(temperature));
    }

    private DateFormat getDateFormat(Date date) {
        return dateUtilsInjector.isToday(date) ? todayDateFormat : normalDateFormat;
    }

    private String capitalize(String string) {
        return string.substring(0, 1).toUpperCase() + string.substring(1).toLowerCase();
    }
}

File: use-cases/src/main/java/com/octo/presentation/CityForecastPresenterImpl.kt
package com.octo.presentation

import com.octo.repository.network.Forecast
import com.octo.usecases.CityForecastPresenter
import java.text.DateFormat
import java.text.DecimalFormat
import java.text.SimpleDateFormat
import java.util.*

class CityForecastPresenterImpl(
    private val view: CityForecastView,
    private val dateUtilsWrapper: DateUtilsWrapper
) : CityForecastPresenter {
    private val normalDateFormat: DateFormat =
        SimpleDateFormat(NORMAL_DATE_FORMAT, Locale.getDefault())
    private val todayDateFormat: DateFormat =
        SimpleDateFormat(TODAY_DATE_FORMAT, Locale.getDefault())

    override fun onEmptyInput() {
        view.displayEmptyInput()
    }

    override fun onGenericException() {
        view.displayGenericException()
    }

    override fun onUnavailableForecasts() {
        view.displayUnavailableForecasts()
    }

    override fun onForecasts(city: String, worstForecast: Forecast, bestForecast: Forecast) {
        view.displayViewModel(
            ForecastViewModel(
                city = city,
                worstTemperatureLabel = getTemperatureLabel(worstForecast.temperature),
                bestTemperatureLabel = getTemperatureLabel(bestForecast.temperature),
                bestForecastLabel = getForecastLabel(worstForecast),
                worstForecastLabel = getForecastLabel(bestForecast)
            )
        )
    }

    private fun getForecastLabel(forecast: Forecast): String {
        val date: Date = forecast.date
        val dateFormatted: String = getDateFormat(date).format(date)
        return capitalize(dateFormatted)
    }

    private fun getTemperatureLabel(temperature: Float): String {
        val decimalFormat = DecimalFormat()
        decimalFormat.maximumFractionDigits = 0
        return decimalFormat.format(temperature)
    }

    private fun getDateFormat(date: Date): DateFormat {
        return if (dateUtilsWrapper.isToday(date)) todayDateFormat else normalDateFormat
    }

    private fun capitalize(string: String): String {
        return string.substring(0, 1).toUpperCase() + string.substring(1).toLowerCase()
    }

    companion object {
        private const val NORMAL_DATE_FORMAT = "EEEE 'à' H'h'"
        private const val TODAY_DATE_FORMAT = "'aujourd''hui à ' H'h'"
    }
}